src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory =
+                options.CacheDirectory ??
+                throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetCachedEnvelopeFiles()
+        {
+            if (!_cacheDirectory.Exists)
+            {
+                return Enumerable.Empty<FileInfo>();
+            }
+
+            return _cacheDirectory.EnumerateFiles("*.envelope");
+        }
+
+        private FileInfo? TryGetNextEnvelopeFile() => GetCachedEnvelopeFiles()
+            .OrderBy(f => f.CreationTimeUtc)
+            .FirstOrDefault();
+
+        private async ValueTask FlushCacheAsync(CancellationToken cancellationToken = default)
+        {
+            while (TryGetNextEnvelopeFile() is { } envelopeFile)
+            {
+                Options.DiagnosticLogger?.LogDebug("Sending cached envelope: {0}", envelopeFile.FullName);
+
+                using var envelope = await Envelope.DeserializeAsync(
+                    envelopeFile.OpenRead(),
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+                envelopeFile.Delete();
+            }
+        }
+
+        private async ValueTask CacheEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
+        {
+            if (GetCachedEnvelopeFiles().Count() >= 50)
+            {
+                return;
+            }
+
+            var envelopeFile = new FileInfo(
+                Path.Combine(
+                    _cacheDirectory.FullName,
+                    $"{DateTimeOffset.Now}_{envelope.TryGetEventId()}_{envelope.GetHashCode()}")
+            );
+
+            _cacheDirectory.Create();
+            using var stream = envelopeFile.Create();
+            await envelope.SerializeAsync(stream, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override async ValueTask ProcessEnvelopeAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            await FlushCacheAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Does it mean we're not storing the `envelope` received here until we deal with cached files?
I believe we should cache the first first of all, to guarantee in the event of a failure (device our of power, process killed, crashed) we didn't lose the event.

The behavior for mobile at least, is to alwauys write to disk first, then try to send. and delete if it succeeded. Also delete it if we got a failure that could happen again once we retry.

Reply:
Good point, will update.